Source/FDWaveformView.swift

```diff
@@ -380,30 +379,57 @@ public class FDWaveformView: UIView {
             let data = NSMutableData(length: readBufferLength)
             CMBlockBufferCopyDataBytes(readBuffer, 0, readBufferLength, data!.mutableBytes)
             CMSampleBufferInvalidate(readSampleBuffer)
-            let sampleCount = readBufferLength / sizeof(Int16)
+
             let samples = UnsafeMutablePointer<Int16>(data!.mutableBytes)
+            let samplesToProcess = readBufferLength / sizeof(Int16)
 
-            for i in 0 ..< sampleCount {
-                guard i % Int(channelCount) == 0 else {
-                    continue // only use the first channel
-                }
-                let rawData = samples[i]
-                let sample = minMaxX(decibel(CGFloat(rawData) / 32768.0), min: noiseFloor, max: 0.0)
-                tally += sample
-                tallyCount += 1
-                if Int(tallyCount) == samplesPerPixel {
-                    let sample = tally / CGFloat(tallyCount)
-                    sampleMax = sampleMax > sample ? sampleMax : sample
-                    outputSamples.append(sample)
-                    tally = 0
-                    tallyCount = 0
-                }
+
+            let filter = [Float](count: samplesPerPixel, repeatedValue: 1.0 / Float(samplesPerPixel))
+            var processingBuffer = [Float](count: samplesToProcess, repeatedValue: 0.0)
+
+            let sampleCount = vDSP_Length(samplesToProcess)
+
+            //Convert 16bit int samples to floats
+            vDSP_vflt16(samples, 1, &processingBuffer, 1, sampleCount)
+
+            //Take the absolute values to get amplitude
+            vDSP_vabs(processingBuffer, 1, &processingBuffer, 1, sampleCount)
+
+            //Convert to dB
+            var zero: Float = 32768.0
+            vDSP_vdbcon(processingBuffer, 1, &zero, &processingBuffer, 1, sampleCount, 1)
+
+            //Clip to [noiseFloor, 0]
+            var ceil: Float = 0.0
+            var noiseFloorFloat = Float(noiseFloor)
+            vDSP_vclip(processingBuffer, 1, &noiseFloorFloat, &ceil, &processingBuffer, 1, sampleCount)
+
+            //Downsample and average
+            let downSampledLength = samplesToProcess / samplesPerPixel
+            var downSampledData = [Float](count: downSampledLength, repeatedValue: 0.0)
+
+            vDSP_desamp(processingBuffer,
+                        vDSP_Stride(samplesPerPixel),
+                        filter, &downSampledData,
+                        vDSP_Length(downSampledLength),
+                        vDSP_Length(samplesPerPixel))
+
+            let range = nextDataOffset..<(nextDataOffset+downSampledLength)
+            var downSampledDataCG = downSampledData.map { (value: Float) -> CGFloat in
+                let element = CGFloat(value)
+                if element > sampleMax { sampleMax = element }
+                return element
             }
+
```

Comment:
You can't cast CGFloat to Float or vice versa, and the accelerate lib works only in Floats, so I had to create the CGFloat values of the downsampled data. Since there is no loop anymore, I figured it would be most efficient to calculate the sample max inside the map.